e2e/v3/jobs3/jobs3.go

```diff
@@ -0,0 +1,433 @@
+	result := make(map[string]Logs)
+
+	// get list of allocs for the job
+	queryOpts := sub.queryOptions()
+	jobsAPI := sub.nomadClient.Jobs()
+	stubs, _, err := jobsAPI.Allocations(sub.jobID, false, queryOpts)
+	must.NoError(sub.t, err, must.Sprintf("failed to query allocations for %s/%s", group, task))
+
+	// get logs for each task in the group allocations
+	for _, stub := range stubs {
+		if stub.TaskGroup == group {
+			result[stub.ID] = sub.getTaskLogs(stub.ID, task)
+		}
+	}
+	return result
+}
+
+func (sub *Submission) getTaskLogs(allocID, task string) Logs {
+	queryOpts := sub.queryOptions()
+	allocAPI := sub.nomadClient.Allocations()
+	alloc, _, err := allocAPI.Info(allocID, queryOpts)
+	must.NoError(sub.t, err, must.Sprintf("failed to query allocation for %s", allocID))
+
+	fsAPI := sub.nomadClient.AllocFS()
+	read := func(path string) string {
+		rc, err := fsAPI.ReadAt(alloc, path, 0, 0, queryOpts)
+		must.NoError(sub.t, err, must.Sprintf("failed to read alloc logs for %s", allocID))
+		b, err := io.ReadAll(rc)
+		must.NoError(sub.t, err, must.Sprintf("failed to read alloc logs for %s", allocID))
+		must.NoError(sub.t, rc.Close(), must.Sprint("failed to close log stream"))
+		return string(b)
+	}
+
+	stdout := fmt.Sprintf("alloc/logs/%s.stdout.0", task)
+	stderr := fmt.Sprintf("alloc/logs/%s.stderr.0", task)
+
+	return Logs{
+		Stdout: read(stdout),
+		Stderr: read(stderr),
+	}
+}
+
+// JobID provides the (possibly) randomized jobID associated with this Submission.
+func (sub *Submission) JobID() string {
+	return sub.jobID
+}
+
+func (sub *Submission) logf(msg string, args ...any) {
+	sub.t.Helper()
+	util3.Log3(sub.t, sub.verbose, msg, args...)
+}
+
+func (sub *Submission) cleanup() {
+	if sub.noCleanup {
+		return
+	}
+
+	// deregister the job that was submitted
+	jobsAPI := sub.nomadClient.Jobs()
+	sub.logf("deregister job %q", sub.jobID)
+	_, _, err := jobsAPI.Deregister(sub.jobID, true, &nomadapi.WriteOptions{
+		Namespace: sub.inNamespace,
+	})
+	test.NoError(sub.t, err, test.Sprintf("failed to deregister job %q", sub.origJobID))
+
+	// force a system gc just in case
+	sysAPI := sub.nomadClient.System()
+	sub.logf("system gc")
+	err = sysAPI.GarbageCollect()
+	test.NoError(sub.t, err, test.Sprint("failed to gc"))
+
+	// todo: should probably loop over the gc until the job is actually gone
+}
+
+type Option func(*Submission)
+
+type Cleanup func()
+
+func Submit(t *testing.T, filename string, opts ...Option) (*Submission, Cleanup) {
+	sub := initialize(t, filename)
+
+	for _, opt := range opts {
+		opt(sub)
+	}
+
+	sub.setClient() // setup base api clients
+	sub.run()       // submit job and wait on deployment
+	sub.waits()     // wait on batch/sysbatch allocations
+
+	return sub, sub.cleanup
+}
+
+func Namespace(name string) Option {
+	return func(sub *Submission) {
+		sub.inNamespace = name
+	}
+}
+func AuthToken(token string) Option {
+	return func(sub *Submission) {
+		sub.authToken = token
+	}
+}
```

Comment:
Another sort of option we may want to consider here in future work is handling intentionally-failing or otherwise degraded job deployments. There's a pattern we use a bunch where we submit a job that's broken and then change something in the cluster to cause it to get re-evaluated and then allocs get created, or the allocs get placed but then are stuck pending on the client until we add something their `template` block needs.